src/config_builder.py

```diff
@@ -143,4 +147,53 @@ def _limits_config(self) -> dict:
             # case of one stream per user.
             "per_stream_rate_limit": f"{self.ingestion_rate_mb}MB",
             "per_stream_rate_limit_burst": f"{self.ingestion_burst_size_mb}MB",
+            # This charmed operator is intended for running a single loki instances, so we don't need to split queries
+            # https://community.grafana.com/t/too-many-outstanding-requests-on-loki-2-7-1/78249/9
+            "split_queries_by_interval": "0",
+        }
+
+    @property
+    def _query_range(self) -> dict:
+        # Ref: https://grafana.com/docs/loki/latest/configure/#query_range
+        return {
+            "parallelise_shardable_queries": False,
+            "results_cache": {
+                "cache": {
+                    "embedded_cache": {
+                        # https://community.grafana.com/t/too-many-outstanding-requests-on-loki-2-7-1/78249/11
+                        "enabled": True
+                    }
+                }
+            },
+        }
+
+    @property
+    def _chunk_store_config(self) -> dict:
+        # Ref: https://grafana.com/docs/loki/latest/configure/#chunk_store_config
+        return {
+            "chunk_cache_config": {
+                "embedded_cache": {
+                    # https://community.grafana.com/t/too-many-outstanding-requests-on-loki-2-7-1/78249/11
+                    "enabled": True
+                }
+            }
+        }
+
+    @property
+    def _frontend(self) -> dict:
+        # Ref: https://grafana.com/docs/loki/latest/configure/#frontend
+        return {
+            # Maximum number of outstanding requests per tenant per frontend; requests beyond this error with HTTP 429.
+            # Default is 2048, but 8cpu16gb can ingest ~3 times more, so set to 4x.
```

Comment:
This does not add up. If it can ingest 3 times more, why are we then multiplying it by 4? 😅

Reply:
8cpu16gb is just what we use for the load test. In the wild this will obviously be sized differently.
I think at this point taking 4x the default would be sufficient.

- Generally the value should vary with vm size because for smaller vm some users may prefer getting 4xx and 5xx for dropped instructions instead of increase in latency.
- At the moment we have "one size fits all" so with this change here I'm effectively proposing to be opinionated favoring latency over failures for smaller VMs, and failures over latency for higher load.

I can delete the comment so it's less confusing :)

Reply:
Tested the same config with 4cpu8gb and 180k loglines/min and it looks good - not much change in resource use compared to before the change.